src/mjolnir/elevationbuilder.cc

```diff
@@ -0,0 +1,233 @@
+      }
+
+      // Edge elevation information. If the edge is forward (with respect to the shape)
+      // use the first value, otherwise use the second.
+      bool forward = directededge.forward();
+      directededge.set_weighted_grade(forward ? std::get<0>(found->second)
+                                              : std::get<1>(found->second));
+      float max_up_slope = forward ? std::get<2>(found->second) : std::get<4>(found->second);
+      float max_down_slope = forward ? std::get<3>(found->second) : std::get<5>(found->second);
+      directededge.set_max_up_slope(max_up_slope);
+      directededge.set_max_down_slope(max_down_slope);
+    }
+
+    // Update the tile
+    tilebuilder.StoreTileData();
+
+    // Check if we need to clear the tile cache
+    if (graphreader.OverCommitted()) {
+      lock.lock();
+      graphreader.Clear();
+      lock.unlock();
+    }
+  }
+}
+
+} // namespace
+
+namespace valhalla {
+namespace mjolnir {
+
+void ElevationBuilder::Build(const boost::property_tree::ptree& pt) {
+
+  auto hierarchy_properties = pt.get_child("mjolnir");
+  std::string tile_dir = hierarchy_properties.get<std::string>("tile_dir");
+
+  // Crack open some elevation data if its there. Return if it is not.
+  boost::optional<std::string> elevation = pt.get_optional<std::string>("additional_data.elevation");
+  std::unique_ptr<const skadi::sample> sample;
+  if (elevation && boost::filesystem::exists(*elevation)) {
+    sample.reset(new skadi::sample(*elevation));
+  } else {
+    LOG_INFO("ElevationBuilder: no elevation data, skipping");
+    return;
+  }
+
+  // Create a randomized queue of tiles (at all levels) to work from
+  std::deque<GraphId> tilequeue;
+  GraphReader reader(pt.get_child("mjolnir"));
+  auto tileset = reader.GetTileSet();
+  for (const auto& id : tileset) {
+    tilequeue.emplace_back(id);
+  }
+  std::random_shuffle(tilequeue.begin(), tilequeue.end());
+
+  // An mutex we can use to do the synchronization
+  std::mutex lock;
+
+  // Setup threads
+  uint32_t nthreads =
+      std::max(static_cast<unsigned int>(1),
+               pt.get<unsigned int>("concurrency", std::thread::hardware_concurrency()));
+  std::vector<std::shared_ptr<std::thread>> threads(nthreads);
+
+  // Setup promises. Hold the results for the threads
+  std::vector<std::promise<uint32_t>> results(nthreads);
+
+  LOG_INFO("Adding elevation to " + std::to_string(tilequeue.size()) + " tiles with " +
+           std::to_string(nthreads) + " threads...");
+
+  // Spawn the threads
+  for (auto& thread : threads) {
+    results.emplace_back();
+    thread.reset(new std::thread(add_elevation, std::cref(pt), std::ref(tilequeue), std::ref(lock),
+                                 std::cref(sample), std::ref(results.back())));
+  }
+
+  // Wait for threads to finish
+  for (auto& thread : threads) {
+    thread->join();
+  }
+
+  /** // Get the promise from the future
+  for (auto& result : results) {
```

Comment:
@dnesbitt61 Should this be removed?  Was this just leftover from testing?

Reply:
Was going to return some sort of stats, any thoughts? If not I guess it can be removed

Reply:
I will create an issue to create some stats for elevation.